### DISCUSSION TITLE
Substitute position

### DISCUSSION THREAD

**Discussion by mameIIas** [pymupdf/PyMuPDF#1737] [Looking for help]
Hi, after searching and replacing a word for another in the text ( the results are amazing), the words seem to be added at the end of the page:
![image](https://user-images.githubusercontent.com/34312469/171128598-4dc61747-9ed4-4ad9-83af-722d78e9fd34.png)
All the replacing words "Eugenia" "Hugo"... after selecting and copy-pasting them, they seem to be at the end, is there any way to keep them in order of appearance?

Thanks in advance!

**Comment by JorjMcKie** [comment]
> is there any way to keep them in order of appearance?

No. You didn't say what you did. I assume you used redaction annots. Anyway, naive copy / paste will never re-establish reading order. Only using Intelligent PDF readers or your own script can do that.

**Reply by mameIIas** [reply]
Yes I did use redaction annots. Is there another way to do that replacing but keeping that order?

**Reply by JorjMcKie** [reply]
No. This is impossible by principle - i.e. not a MuPDF weakness or whatever.
The only way is to completely rewrite all the text of the page.
Or do a text extraction intelligent enough to reorder the words by their geometrical position info. For example, use `page.get_text("words", sort=True)` and then concatenate the words such that each line consists of those having the same bottom y-coordinate.

**Reply by mameIIas** [reply]
Ookay, then it'd be this replacing, save it on a PDF and then run the get_text? Or even just after the replacing?

**Reply by JorjMcKie** [reply]
just after replacing should already work too

**Reply by mameIIas** [reply]
Okay, thank you so much!

**Reply by mameIIas** [reply]
Sorry for the dumb question, but how do I concatenate the words:

```
page.addRedactAnnot(rect)  # redact the word
page.apply_redactions()  # and imediately apply!
page.insertText(origin, sustituto, fontname=font, fontsize=fsize)

words = page.get_text("words", sort=True)
i = 0

while i < len(words):
    word = words[i]
    word2 = words[i+1]
    if math.isclose (word[1], word2[1]):  #  check if same line
        #concatenate word

        i += 1
    else:
    #new line
    #concatenate
        i += 1

doc.save(ouput)
```

And that would be that structure?
Thanks in advance!

**Reply by JorjMcKie** [reply]
SOmething like that ...
But remember: this only extracts text in reading sequence - it does not change the PDF of course!

```python
words=page.get_text("words",sort=True)
text=words[0][4]  # start a line with the first word text
for i,w in enumerate(words[1:],start=1):
    if abs(w[3]-words[i-1][3])<=2:  # if bottom coord are close enough
        text += " " + w[4]  # append text of next word
    else:  # output line text
        print(text)
        text=w[4]  # and start new line text
print(text)  # print last line
```

**Reply by mameIIas** [reply]
Just tested that and prints perfectly the text after that replacing. The PDF looks just how I wanted it to look, but whhen selecting the substitute is still at the end, my solution now would be re-writing The PDF with that sequence?

**Reply by JorjMcKie** [reply]
> ...  my solution now would be re-writing The PDF with that sequence?

Can't imagine why you would absolutely want to do that. First it is not trivial, and second many, **_many_** original PDFs contain similar corrections ...

**Reply by mameIIas** [reply]
Just to try to solve this:
![image](https://user-images.githubusercontent.com/34312469/171179228-60bff3ed-340f-42ab-a2ae-22edd5a19625.png)
What stated before about selecting and copy pasting, and thought that with that kind of sort that would be achievable

**Reply by mameIIas** [reply]
You said that "s. Anyway, naive copy / paste will never re-establish reading order. Only using Intelligent PDF readers or your own script can do that." So that's the only way I have to try to solve that? And any advice how to approach that idea?

Thanks in advance!

**Reply by JorjMcKie** [reply]
As I wrote: not trivial! The basic approach is:
1. Extract all text from the page via `page.get_text("dict",...)`.
2. Make sure you have available all fonts used by the page's text on your computer.
3. Remove **all text** from the page. This is a hacky procedure as done in the font replacement scripts in the utilities repo.
4. Re-insert the text - modified in the way you want (replacing certain words or whatever).

Or simply use a tool that can directly edit a PDF. E.g. MS Word does work for a number of PDFs.
But I know no tool that is perfect.

**Reply by JorjMcKie** [reply]
Also OpenOffice and LibreOffice

**Reply by mameIIas** [reply]
I'm going the not-trivial way, I found that clean_fontnames and do execute it, but tells me not fonts to subset. It only has 3, and both are installed, they are listed in fontlist, but not through fontref. What can I do? I'm stuck in phase 1 of that replacement script because "max() arg is an empty sequence". It is not getting that fontnames and I can't get it to work.

Sorry for all this questions

**Reply by JorjMcKie** [reply]
Too busy with paid work to help, sorry.
Good luck anyway.

**Reply by mameIIas** [reply]
Wow, thanks anyway.

**Reply by mameIIas** [reply]
Hi again, I hope you have a bit of time to answer this question. I'm trying to replace Helvetica for Helvetica, so it's the same font, but the order will be fixed. I'm trying to pass it directly, but keeps me getting error because I don't know how to define it:
![image](https://user-images.githubusercontent.com/34312469/173020065-95d2df3d-bea9-45c5-b5e6-663ab53aa30b.png)

```fontrefs` = {b"/Helvetica": 'Helvetica'}```

```new_fontnames = {'Helvetica':'Helvetica'}```

Or how do I pass that name?

Thanks in advance again, hope you can help me

**Reply by mameIIas** [reply]
I finally do get it to run but with error:
![image](https://user-images.githubusercontent.com/34312469/173525392-4656dc5a-2c6e-472f-a506-2f3364a42667.png)
I don't know how to declare these two, that causes the next error:
![image](https://user-images.githubusercontent.com/34312469/173525574-aad5bf88-26cd-449f-8b3b-d1bba89e2999.png)
At least now I get the fontrefs.
And after running that program with the pdf replaced before, it only deletes the replaced part:
![image](https://user-images.githubusercontent.com/34312469/173525895-88734d55-4474-4ea7-b95b-abc1f8d5cc76.png)
Any clue to get that to work, to get that selected part in order.
Thanks in advance
